core/corehttp/gateway_handler.go

```diff
@@ -112,10 +123,17 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	w.Header().Set("X-IPFS-Path", urlPath)
 
 	// Suborigin header, sandboxes apps from each other in the browser (even
-	// though they are served from the same gateway domain). NOTE: This is not
-	// yet widely supported by browsers.
-	pathRoot := strings.SplitN(urlPath, "/", 4)[2]
-	w.Header().Set("Suborigin", pathRoot)
+	// though they are served from the same gateway domain).
+	//
+	// Omited if the path was treated by IPNSHostnameOption(), for example
+	// a request for http://example.net/ would be changed to /ipns/example.net/,
+	// which would turn into an incorrect Suborigin: example.net header.
```

Comment:
With IPNSHostname the gateway serves one object per hostname, while otherwise it serves all the objects that exist (through /ipfs and /ipns). I thought there'd be no need to separate it from anything, while under /ipfs and /ipns, each object and identity should be separated.